fl4health/losses/mkmmd_loss.py

```diff
@@ -167,12 +216,50 @@ def forward(self, Xs: torch.Tensor, Xt: torch.Tensor) -> torch.Tensor:
         #   G is (kernel_num, kernel_num)
         #   h is (kernel_num, 1)
         #   A is (1, kernel_num)
-        #   b is ()
-        beta = self.sign * QPFunction(verbose=False)(Q, p, G, h, A, b)
+        #   b is a scalar
+
+        # QPFunction returns betas in the shape 1 x num_kernels to we take the transpose for consistency
+        return QPFunction(verbose=False, solver=QPSolvers.CVXPY)(regularized_Q_k, p, G, h, hat_d_per_kernel.t(), b).t()
+
+    def optimize_betas(self, X: torch.Tensor, Y: torch.Tensor, lambda_m: float) -> torch.Tensor:
+        # In this function, we assume that X, Y: n_samples, n_features
+        all_h_u_per_v_i = self.compute_all_h_u_per_v_i(X, Y)
+
+        # shape of hat_d_per_kernel is number of kernels x 1
+        hat_d_per_kernel = self.compute_hat_d_per_kernel(all_h_u_per_v_i)
+        # check to see that at least one of them is positive. If none of them are positive, then select a single kernel
+        # with largest hat_d, similar to the suggestion of Gretton et al. in "Optimal Kernel Choice for Large-Scale
+        # Two-Sample Tests", 2012
+        if not torch.any(hat_d_per_kernel > 0):
+            log(
+                INFO,
+                f"None of the estimates for hat_d are positive: {hat_d_per_kernel}. Rather than optimizing, we select "
+                "a single kernel with largest hat_d_k",
+            )
+            return self.beta_with_largest_hat_d(hat_d_per_kernel)
```

Comment:
shouldn't we return beta with smallest hat d if we want to maximize the distance between two distributions? (minimize_type_two_error=False)

Reply:
or maybe we should first the condition `if self.minimize_type_two_error` ?

Reply:
Yes, you're right. We should only return this if we're minimizing the type two error. Good call.

Reply:
Otherwise, we're still good to make it to the else statement below.